### ISSUE TITLE
Map prototype collision?

### ISSUE THREAD

**Issue by shamilovtim** [LegendApp/legend-state#118]
I was thinking about how both `Map` and legend-state have methods named `get()`. How is this collision handled if putting a `Map` inside a legend state atom?

**Comment by jmeistrich** [comment]
I haven't really come up with the right way to do collections like Set and Map. It would be ideal if we could make them observable like other objects, but I haven't gotten to thinking through that yet. So in general I would say probably avoid using them in observables for now.

But to answer the actual question, I'm not sure what it should do. I like `get()` as a good name for what it does. But it is a problem that it conflicts with `Map` and `Set`. 

What do you think it should do?

**Comment by shamilovtim** [comment]
what I would do is push a breaking change which removes this collision by renaming `get()` and `set()`. rename `get` to `track` and `set` to `write` and any other injected methods that conflict with similar reserved words or APIs

**Comment by jmeistrich** [comment]
@shamilovtim Version 1.2.0 adds support for adding Map and Set to observables. I went with a strategy of overloading the functions to call the observable function vs. the Map function based on the number of arguments. So:

`get()` calls the Observable function
`get("key")` calls the Map function
`set(map)` calls the Observable function
`set(key, value)` calls the Map function
`delete()` calls the Observable function
`delete(key)` calls the Map function

This seems to work pretty well, and I massaged the TypeScript types so that it presents the correct overloads. What do you think?

**Comment by shamilovtim** [comment]
@jmeistrich Very cool! A question. Valtio has said that its implementation of Map using Proxy was an emulation and much costlier than just using Arrays and Objects, so any benefits of Map or Set were negated. Is that also the case for this Map / Set implementation? Given the idea that Proxy doesn't natively support Map/Set is there any valid performance reason for using Map / Set here or would the only case be for code legibility or the other soft benefits of Map/Set?

(I am aware that for the differences between Map and Object/Array to be relevant there has to be a lot of data but on React Native these types of differences emerge quickly in contradiction to what some JS gurus say either because they ignore RN or they aren't aware of tuning RN to keep the JS thread unblocked)

**Comment by jmeistrich** [comment]
There is no performance cost. Legend-State uses Proxy differently than Valtio - it proxies around the path to each node instead of the data itself, and doesn't touch the underlying data at all, so it doesn't need to change the normal Map behavior. So if you have

```js
const obs = observable({ chat: { messages: new Map(['m1', { text: 'hello' }]) } })
const message$ = obs.chat.messages.get('m1')
```

The proxy for `message$` observable is just a reference to `['chat', 'messages', 'm1']` within the object. So to get the data it just traverses the object from the root down the path, through the Map's normal `get(key)`.

**Comment by shamilovtim** [comment]
So I take it For would natively support this then? Very cool

**Comment by jmeistrich** [comment]
Haha I'm working on that now :)

**Comment by jmeistrich** [comment]
I'm going to close this as completed now. Please let me know if it can still be improved further. Thanks!

**Comment by olibos** [comment]
Hello @jmeistrich ,

I'm trying to use this on my project but I have some issues:
```ts
const products = observable(new Map<string, number>());
products.get('id') // -> ObservableBaseFns<T>.get(key: string): number | undefined
```

it seems that in typescript the get with key returns the value instead of an observable, but at runtime I receive the proxy.
It looks like a typing issue.

Another thing is when I try to get notified on size change with:
```ts
products.size.use()
```
It always returns undefined and there is no re-render on item changes.

Maybe not the correct usage.

**Comment by jmeistrich** [comment]
@olibos Ah good catch, thanks! It was only typing Map correctly if it was a child, not if it was at the root of the observable. And it was not special casing "size" to be an observable.

This should be fixed in version 1.10.1.
